src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -305,21 +306,22 @@ private void saveModelCheckpointAsync(Map<String, Object> source, String modelId
     public Map<String, Object> toIndexSource(ModelState<EntityModel> modelState) throws IOException {
         Map<String, Object> source = new HashMap<>();
         EntityModel model = modelState.getModel();
-        String serializedModel = toCheckpoint(model);
-        if (serializedModel == null || serializedModel.length() > maxCheckpointBytes) {
+        Optional<String> serializedModel = toCheckpoint(model, modelState.getModelId());
+        if (!serializedModel.isPresent() || serializedModel.get().length() > maxCheckpointBytes) {
             logger
                 .warn(
                     new ParameterizedMessage(
-                        "[{}]'s model empty or too large: [{}] bytes",
+                        "[{}]'s model is empty or too large: [{}] bytes",
                         modelState.getModelId(),
-                        serializedModel == null ? 0 : serializedModel.length()
+                        serializedModel.isPresent() ? serializedModel.get().length() : 0
                     )
                 );
             return source;
         }
         String detectorId = modelState.getDetectorId();
         source.put(DETECTOR_ID, detectorId);
-        source.put(FIELD_MODELV2, serializedModel);
+        // we cannot pass Optional as OpenSearch does not know how to serialize an Optional value
+        source.put(FIELD_MODELV2, serializedModel.get());
```

Comment:
when an optional is empty, get returns an exception.

Reply:
yes, we returned early if that the optional is empty.  Check line 310~320.

Reply:
ok. that's good.